In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/merged-tinkoff/test_data.csv
/kaggle/input/merged-tinkoff/train_data.csv
/kaggle/input/merged-tinkoff/new_stories.csv


https://colab.research.google.com/drive/1OhPnSdUvAyb1QjIaWpJtVSHVO4_cvFvP here is preprocessing of stories, used colab because needed Python 3.6 to make Dostoevsky work.

In [3]:
import pandas as pd
import numpy as np

In [4]:
#loading data
train_data = pd.read_csv('/kaggle/input/merged-tinkoff/train_data.csv')
stories = pd.read_csv('/kaggle/input/merged-tinkoff/new_stories.csv')

In [5]:
#checking the data
train_data

customer_id  story_id           event_dttm  event  month  weekday  \
0                15       138  2018-07-24 15:33:22      3      7        1   
1                15       202  2018-06-04 08:08:08      2      6        0   
2                15       222  2018-06-17 13:44:45      2      6        6   
3                15       379  2018-05-23 05:41:43      2      5        2   
4                15       544  2018-07-25 02:16:29      3      7        2   
...             ...       ...                  ...    ...    ...      ...   
448001      1103495      1220  2018-07-16 06:26:35      2      7        0   
448002      1103495      1234  2018-07-15 19:44:38      3      7        6   
448003      1103495      1234  2018-07-16 06:26:35      2      7        0   
448004      1103495      1264  2018-07-15 19:44:38      3      7        6   
448005      1103495      1264  2018-07-16 06:26:35      2      7        0   

        day  hour  minute  is_weekend  ...  No info_amt  Pharmacy_amt  \
0        24    15      33           0  ...      1580500          8000   
1         4     8       8           0  ...      1580500          8000   
2        17    13      44           1  ...      1580500          8000   
3        23     5      41           0  ...      1580500          8000   
4        25     2      16           0  ...      1580500          8000   
...     ...   ...     ...         ...  ...          ...           ...   
448001   16     6      26           0  ...        35000             0   
448002   15    19      44           1  ...        35000             0   
448003   16     6      26           0  ...        35000             0   
448004   15    19      44           1  ...        35000             0   
448005   16     6      26           0  ...        35000             0   

        Photo and Video_amt  Restaurants_amt  Souvenirs_amt  \
0                         0            27500           2500   
1                         0            27500           2500   
2                         0            27500           2500   
3                         0            27500           2500   
4                         0            27500           2500   
...                     ...              ...            ...   
448001                    0                0              0   
448002                    0                0              0   
448003                    0                0              0   
448004                    0                0              0   
448005                    0                0              0   

        Sporting goods_amt  Supermarkets_amt  Train Tickets_amt  \
0                    33000            198500                  0   
1                    33000            198500                  0   
2                    33000            198500                  0   
3                    33000            198500                  0   
4                    33000            198500                  0   
...                    ...               ...                ...   
448001                   0             12500                  0   
448002                   0             12500                  0   
448003                   0             12500                  0   
448004                   0             12500                  0   
448005                   0             12500                  0   

        Transport_amt  Utility Services_amt  
0               15000                  2500  
1               15000                  2500  
2               15000                  2500  
3               15000                  2500  
4               15000                  2500  
...               ...                   ...  
448001              0                     0  
448002              0                     0  
448003              0                     0  
448004              0                     0  
448005              0                     0  

[448006 rows x 54 columns]

In [6]:
#train_data.isnull().sum()

In [7]:
#filling missing values in product coluns using category 4 wich corresponds to unknown values
for i in range(7):
        train_data[f'product_{i}'] = train_data[f'product_{i}'].fillna(4)

In [8]:
#merging both datasets into one
full_data = train_data.merge(stories, on='story_id')

In [9]:
full_data

customer_id  story_id           event_dttm  event  month  weekday  \
0                15       138  2018-07-24 15:33:22      3      7        1   
1              2846       138  2018-05-29 04:01:17      3      5        1   
2              3456       138  2018-05-25 16:58:51      2      5        4   
3              3618       138  2018-07-20 09:53:27      3      7        4   
4              3756       138  2018-06-17 03:39:20      2      6        6   
...             ...       ...                  ...    ...    ...      ...   
445009       729421      1213  2018-07-28 04:54:47      3      7        5   
445010       714105      1405  2018-07-31 18:40:21      2      7        1   
445011       843538       609  2018-04-03 05:13:38      2      4        1   
445012       843538       625  2018-04-03 05:13:38      2      4        1   
445013       917541       656  2018-06-19 05:17:20      2      6        1   

        day  hour  minute  is_weekend  ...  Utility Services_amt  \
0        24    15      33           0  ...                  2500   
1        29     4       1           0  ...                  1500   
2        25    16      58           0  ...                  2000   
3        20     9      53           0  ...                     0   
4        17     3      39           1  ...                  2000   
...     ...   ...     ...         ...  ...                   ...   
445009   28     4      54           1  ...                  4500   
445010   31    18      40           0  ...                     0   
445011    3     5      13           0  ...                     0   
445012    3     5      13           0  ...                     0   
445013   19     5      17           0  ...                     0   

        story_length  slides_count  text_len  name_len   neutral  negative  \
0                  3             3       185        15  0.948165  0.067557   
1                  3             3       185        15  0.948165  0.067557   
2                  3             3       185        15  0.948165  0.067557   
3                  3             3       185        15  0.948165  0.067557   
4                  3             3       185        15  0.948165  0.067557   
...              ...           ...       ...       ...       ...       ...   
445009             9            25      2636        29  0.637041  0.136618   
445010             3             6       142        17  0.859674  0.073706   
445011             1             5       171         9  0.743178  0.191943   
445012             8            37      2267        19  0.554480  0.132974   
445013             9            10       855        25  0.939923  0.119213   

            skip    speech  positive  
0       0.122533  0.005565  0.011342  
1       0.122533  0.005565  0.011342  
2       0.122533  0.005565  0.011342  
3       0.122533  0.005565  0.011342  
4       0.122533  0.005565  0.011342  
...          ...       ...       ...  
445009  0.112805  0.023699  0.095359  
445010  0.112805  0.002725  0.073706  
445011  0.281416  0.009423  0.009718  
445012  0.103759  0.016413  0.168867  
445013  0.053413  0.006300  0.022987  

[445014 rows x 63 columns]

In [10]:
train_data.shape, full_data.shape

((448006, 54), (445014, 63))

In [11]:
#full_data.isnull().sum().tail(50)

In [12]:
#filling categorical and one numerical columns with '-1', age - with median age
#dropping unused columns
columns_to_neg = ['marital_status_cd', 'children_cnt', 'job_position_cd', 'gender_cd', 
                  'months_after_first_session']
cols_drop = ['first_ses', 'first_session_dttm', 'event_dttm',
            'minute', 'is_weekend', 'weekday', 'month', 'positive', 'skip']
full_data[columns_to_neg] = full_data[columns_to_neg].fillna(-1)
full_data['age']=full_data['age'].fillna(full_data['age'].median())
full_data = full_data.drop(cols_drop, axis=1)

In [13]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

In [14]:
#splitting dataset to train and validation datasets
X, y = full_data.drop(columns=['event']), full_data['event']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=.4, random_state=10)
y_train.value_counts()

3    137745
2     93140
4     32387
1      3736
Name: event, dtype: int64

0.85% of common customer_id are in X_val
for train and test sets it is 0.59%

In [15]:
test_data = pd.read_csv('/kaggle/input/merged-tinkoff/test_data.csv')
cus_tr = train_data.customer_id.unique().tolist()
cus_tes=test_data.customer_id.unique().tolist()
common = [i for i in cus_tr if i in cus_tes]
print('For train and test sets:', len(common), len(cus_tr), len(cus_tes), len(common)/len(cus_tes),"%")


c_train = X_train.customer_id.unique().tolist()
c_val= X_val.customer_id.unique().tolist()
comm = [i for i in c_train if i in c_val]
print('For X_train and X_valid sets:', len(comm), len(c_train), len(c_val), len(comm)/len(c_val), '%')


For train and test sets: 13463 36446 22646 0.594497924578292 %
For X_train and X_valid sets: 24065 32089 28390 0.8476576259246213 %


In [16]:
#adding target columns to both datasets because we need them to do some processing on the train data
#excluding validation data
X_train['target'] = y_train
X_val['target'] = y_val 

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [17]:
#parsing the labels, getting how many dislikes, skips, views and likes are there per user
sparse = pd.pivot_table(X_train, values = ['story_id'],
                                     index=['customer_id'], aggfunc='count',
                     columns=['target'], fill_value = 0).add_suffix('_user_gave')
sparse.head(10)

story_id_user_gave                                    
target             1_user_gave 2_user_gave 3_user_gave 4_user_gave
customer_id                                                       
15                           0           6          23           0
73                           0           7           1           0
91                           0           0           1           0
178                          0           4           0           0
362                          1           0          10           0
521                          0          21         129           0
536                          0           1           4           3
612                          0           1           5           0
629                          0           0           3           1
791                          0           1           1           0

In [18]:
#getting how many dislikes, skips, views and likes each story got
sparse2 = pd.pivot_table(X_train, values = ['customer_id'],
                                     index=['story_id'], aggfunc='count',
                     columns=['target'], fill_value = 0).add_suffix('_story_got')
sparse2.head(10)

customer_id_story_got                                    
target             1_story_got 2_story_got 3_story_got 4_story_got
story_id                                                          
123                          0         193         313           0
126                          0         347         103           0
127                          0          32          15           0
128                          0         242         308           3
129                          0         116         105           0
132                          0          85          78           0
133                          0          49          36           0
134                          0          27          31           0
135                          0         117         138           0
136                          1         111          43           0

In [19]:
#merging datasets and dropping the targets
X_train = X_train.merge(sparse, on=['customer_id'])
X_train = X_train.merge(sparse2, on=['story_id'])

X_val = X_val.merge(sparse, on=['customer_id'])
X_val = X_val.merge(sparse2, on=['story_id'])

y_train = X_train['target']
y_val = X_val['target']
X_train.drop(['target'], axis=1, inplace=True)
X_val.drop(['target'], axis=1, inplace=True)

/opt/conda/lib/python3.7/site-packages/pandas/core/reshape/merge.py:643: UserWarning: merging between different levels can give an unintended result (1 levels on the left,2 on the right)
  warnings.warn(msg, UserWarning)


In [20]:
X_train

customer_id  story_id  day  hour  story_count  product_0  product_1  \
0             43943      1344   23     3           92        4.0        2.0   
1            370840      1344   20     4          190        4.0        4.0   
2            212489      1344   21     4          155        4.0        4.0   
3            209288      1344   20     4           87        4.0        4.0   
4            580878      1344   18     7          262        4.0        4.0   
...             ...       ...  ...   ...          ...        ...        ...   
267003        61621      1335   22    10            9        4.0        4.0   
267004        37361       831   11     5            2        4.0        4.0   
267005       582178       831   11     4            1        4.0        2.0   
267006       145538       831   10    12            3        4.0        4.0   
267007       148756      1333   31     4            8        4.0        4.0   

        product_2  product_3  product_4  ...  negative    speech  \
0             2.0        2.0        4.0  ...  0.136618  0.006914   
1             2.0        3.0        4.0  ...  0.136618  0.006914   
2             4.0        3.0        4.0  ...  0.136618  0.006914   
3             1.0        2.0        4.0  ...  0.136618  0.006914   
4             2.0        3.0        4.0  ...  0.136618  0.006914   
...           ...        ...        ...  ...       ...       ...   
267003        4.0        2.0        4.0  ...  0.125933  0.024433   
267004        4.0        2.0        4.0  ...  0.222710  0.009136   
267005        4.0        4.0        4.0  ...  0.222710  0.009136   
267006        4.0        4.0        4.0  ...  0.222710  0.009136   
267007        4.0        2.0        4.0  ...  0.140346  0.008587   

        (story_id_user_gave, 1_user_gave)  (story_id_user_gave, 2_user_gave)  \
0                                       0                                 12   
1                                       1                                 24   
2                                       0                                 24   
3                                       1                                 18   
4                                       4                                 48   
...                                   ...                                ...   
267003                                  0                                  1   
267004                                  0                                  0   
267005                                  0                                  1   
267006                                  0                                  0   
267007                                  0                                  2   

        (story_id_user_gave, 3_user_gave)  (story_id_user_gave, 4_user_gave)  \
0                                      16                                 32   
1                                      92                                  5   
2                                      76                                  0   
3                                      17                                 13   
4                                      65                                 33   
...                                   ...                                ...   
267003                                  1                                  2   
267004                                  1                                  0   
267005                                  0                                  0   
267006                                  1                                  0   
267007                                  1                                  0   

        (customer_id_story_got, 1_story_got)  \
0                                          1   
1                                          1   
2                                          1   
3                                          1   
4                                          1   
...                             

In [21]:
X_train.shape, X_val.shape, y_train.shape, y_val.shape

((267008, 61), (171963, 61), (267008,), (171963,))

In [22]:
#refreshing of label values and weights function
def label_to_id():
    return {
        'dislike': 1,
        'skip': 2, 
        'view': 3,
        'like': 4,
    }

def id_to_w():
    return {
        1: -10,
        2: -0.1,
        3: 0.1,
        4: 0.5,
    }

#writing a function for evaluating the results. 
#probs stands for probablitiies output of a model
#we multiply outputs with real weights and then divide this score on real weights
def score(probs, y_true):
    answers = probs
    id2w = id_to_w()
    true_weights = np.array(list(map(id2w.get, y_true)))
    my_score = (true_weights * answers).sum()
    true_score = np.abs(true_weights).sum()
    
    return my_score / true_score, my_score, true_score

In [23]:
#basic RFC model, tuned by hand, checked max_depth till 8, but dropping this parameter gave better results
#n_estimators- checked 200, 300, 500, 1000
model = RandomForestClassifier(n_estimators=500, random_state=42)
model.fit(X_train, y_train)

val_probs = model.predict_proba(X_val)
weights = np.array([[-10, -0.1, 0.1, 0.5]])
weighted_ans = np.sum(val_probs * weights, axis=1)

score(weighted_ans, y_val)

(0.7259666723267356, 36674.5295459365, 50518.200000000004)

In [24]:
val_probs

array([[0.004, 0.354, 0.614, 0.028],
       [0.09 , 0.324, 0.07 , 0.516],
       [0.002, 0.12 , 0.758, 0.12 ],
       ...,
       [0.004, 0.138, 0.738, 0.12 ],
       [0.002, 0.62 , 0.31 , 0.068],
       [0.004, 0.892, 0.07 , 0.034]])

In [25]:
from xgboost import XGBClassifier
model_prob = XGBClassifier(n_estimators= 500, random_state=42)
model_prob.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=500, n_jobs=0, num_parallel_tree=1,
              objective='multi:softprob', random_state=42, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=None, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [26]:
v_probs = model_prob.predict_proba(X_val)
weighted_ans_xgb = np.sum(v_probs * weights, axis=1)
score(weighted_ans_xgb, y_val)

(0.8166836177265155, 41257.38633703166, 50518.200000000004)

In [27]:
v_probs

array([[2.0638167e-06, 3.8265803e-01, 6.1733508e-01, 4.8485149e-06],
       [4.6518777e-02, 6.0547405e-01, 8.2864635e-02, 2.6514253e-01],
       [3.0154263e-07, 9.5888764e-02, 8.0836964e-01, 9.5741279e-02],
       ...,
       [1.4486216e-07, 1.6553019e-04, 9.9979883e-01, 3.5492976e-05],
       [3.6992724e-06, 9.7366852e-01, 2.6200522e-02, 1.2720654e-04],
       [1.0569153e-06, 9.9999297e-01, 4.8058009e-06, 1.1306057e-06]],
      dtype=float32)

In [28]:
#feature importance for RFC
fi = pd.DataFrame({'feature': list(X_train),
                   'importance': model.feature_importances_}).\
                    sort_values('importance', ascending = False)
print(fi.importance.sum())
fi.head(50)

1.0


feature  importance
55     (story_id_user_gave, 3_user_gave)    0.089771
54     (story_id_user_gave, 2_user_gave)    0.070874
56     (story_id_user_gave, 4_user_gave)    0.068602
4                            story_count    0.038509
60  (customer_id_story_got, 4_story_got)    0.033270
3                                   hour    0.032556
2                                    day    0.031971
58  (customer_id_story_got, 2_story_got)    0.030215
59  (customer_id_story_got, 3_story_got)    0.029166
48                              text_len    0.028283
1                               story_id    0.028191
50                               neutral    0.026980
51                              negative    0.025572
57  (customer_id_story_got, 1_story_got)    0.025001
52                                speech    0.024537
47                          slides_count    0.021941
49                              name_len    0.020898
0                            customer_id    0.018473
20                   num_transaction_amt    0.018302
19                  mean_transaction_amt    0.018058
21                  std_transactions_amt    0.017991
18                   sum_transaction_amt    0.017358
17            months_after_first_session    0.017149
42                      Supermarkets_amt    0.016496
46                          story_length    0.016205
36                           No info_amt    0.015290
31                          Fastfood_amt    0.013435
28                Clothing and shoes_amt    0.012603
39                       Restaurants_amt    0.012141
53     (story_id_user_gave, 1_user_gave)    0.012089
34                              Fuel_amt    0.011003
13                                   age    0.010716
37                          Pharmacy_amt    0.010684
35                     House, Repair_amt    0.010579
44                         Transport_amt    0.010126
25                            Beauty_amt    0.008705
45                  Utility Services_amt    0.008362
16                       job_position_cd    0.008123
30                    Entertainments_amt    0.007182
26                 Business Services_amt    0.007090
41                    Sporting goods_amt    0.006903
14                     marital_status_cd    0.006585
24                      Auto Service_amt    0.006509
23                Art, Cinema, Music_amt    0.005990
43                     Train Tickets_amt    0.005264
33                           Flowers_amt    0.004710
15                          children_cnt    0.004086
6                              product_1    0.004015
12                             gender_cd    0.003827
8                              product_3    0.003726

In [29]:
#feature importance for XGBoost
fi_xgb = pd.DataFrame({'feature': list(X_train),
                  'importance': model_prob.feature_importances_}).\
                   sort_values('importance', ascending = False)
print(fi.importance.sum())
fi_xgb.head(50)

1.0


feature  importance
56     (story_id_user_gave, 4_user_gave)    0.190910
55     (story_id_user_gave, 3_user_gave)    0.153480
54     (story_id_user_gave, 2_user_gave)    0.114450
53     (story_id_user_gave, 1_user_gave)    0.073092
60  (customer_id_story_got, 4_story_got)    0.044349
4                            story_count    0.022952
58  (customer_id_story_got, 2_story_got)    0.017571
57  (customer_id_story_got, 1_story_got)    0.017154
59  (customer_id_story_got, 3_story_got)    0.016224
48                              text_len    0.010977
50                               neutral    0.008639
47                          slides_count    0.008626
1                               story_id    0.008424
46                          story_length    0.008354
13                                   age    0.007917
51                              negative    0.007902
12                             gender_cd    0.007880
5                              product_0    0.007393
2                                    day    0.007292
31                          Fastfood_amt    0.006914
25                            Beauty_amt    0.006781
36                           No info_amt    0.006700
42                      Supermarkets_amt    0.006664
34                              Fuel_amt    0.006640
27                        Car Rental_amt    0.006612
45                  Utility Services_amt    0.006599
41                    Sporting goods_amt    0.006598
39                       Restaurants_amt    0.006597
21                  std_transactions_amt    0.006572
20                   num_transaction_amt    0.006548
30                    Entertainments_amt    0.006520
0                            customer_id    0.006510
18                   sum_transaction_amt    0.006480
17            months_after_first_session    0.006456
44                         Transport_amt    0.006442
28                Clothing and shoes_amt    0.006439
37                          Pharmacy_amt    0.006427
29                         Duty Free_amt    0.006422
22                           Animals_amt    0.006421
35                     House, Repair_amt    0.006413
32                          Finances_amt    0.006360
24                      Auto Service_amt    0.006349
26                 Business Services_amt    0.006340
16                       job_position_cd    0.006303
8                              product_3    0.006292
43                     Train Tickets_amt    0.006286
40                         Souvenirs_amt    0.006280
19                  mean_transaction_amt    0.006272
6                              product_1    0.006260
52                                speech    0.006237

In [30]:
#for val_test = 40%
#this confusion matrix with the sparsed features of likes|dislikes per stories and per user gave 
#the best recall(0.19, the previous ones were 0.11-0.13) for dislikes as well precision is high enoough.
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
pred_classical = model.predict(X_val)
print(classification_report(y_val, pred_classical))
print(confusion_matrix(y_val, pred_classical))

              precision    recall  f1-score   support

           1       0.59      0.16      0.26      2493
           2       0.62      0.53      0.57     59948
           3       0.67      0.77      0.72     87919
           4       0.63      0.55      0.58     21603

    accuracy                           0.65    171963
   macro avg       0.63      0.50      0.53    171963
weighted avg       0.65      0.65      0.64    171963

[[  407   499   914   673]
 [   78 31781 25588  2501]
 [   70 16082 67887  3880]
 [  134  2654  7041 11774]]


In [31]:
#for val_test = 20%
#here precision is even higher
#the most interesting thing is that the accuracy did not improve with using sparsed features, 
#but feature importance tables proved importance of those features and the score improved drastically:
#from 74 to 82.7 for RFC and from 67 to 83.1 for XGBoost if we use val_set as 20% of the full_data.

pred_clas = model_prob.predict(X_val)
print(classification_report(y_val, pred_clas))
print(confusion_matrix(y_val, pred_clas))

              precision    recall  f1-score   support

           1       0.40      0.20      0.26      2493
           2       0.63      0.55      0.58     59948
           3       0.68      0.76      0.72     87919
           4       0.59      0.56      0.57     21603

    accuracy                           0.65    171963
   macro avg       0.57      0.52      0.53    171963
weighted avg       0.65      0.65      0.65    171963

[[  487   467   808   731]
 [  194 32848 23845  3061]
 [  235 16455 66389  4840]
 [  299  2760  6373 12171]]


In [32]:
y_val.head()

0    3
1    2
2    3
3    4
4    4
Name: target, dtype: int64

In [33]:
def transform_pr1(predictions): 
    good = predictions[:,2] + predictions[:,3]
    bad = predictions[:,0] + predictions[:,1]
    result = good - bad
  
    result[result >= 0] = 1
    result[result < 0] = -1
    return result

def transform_labels(label):
    if label==3 or label==4:
        return 1
    if label==1 or label==2:
        return -1
y_val_new = y_val.apply(lambda label : transform_labels(label))

In [34]:
print(y_val_new)

0         1
1        -1
2         1
3         1
4         1
         ..
171958   -1
171959    1
171960    1
171961    1
171962   -1
Name: target, Length: 171963, dtype: int64


In [263]:
transformed_1 = transform_pr1(val_probs)
print(transformed_1)

[ 1.  1.  1. ...  1. -1. -1.]


In [36]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_val_new, transformed_1)) #for RF this score is 72,01

0.7225624116815826


In [37]:
v_probs

array([[2.0638167e-06, 3.8265803e-01, 6.1733508e-01, 4.8485149e-06],
       [4.6518777e-02, 6.0547405e-01, 8.2864635e-02, 2.6514253e-01],
       [3.0154263e-07, 9.5888764e-02, 8.0836964e-01, 9.5741279e-02],
       ...,
       [1.4486216e-07, 1.6553019e-04, 9.9979883e-01, 3.5492976e-05],
       [3.6992724e-06, 9.7366852e-01, 2.6200522e-02, 1.2720654e-04],
       [1.0569153e-06, 9.9999297e-01, 4.8058009e-06, 1.1306057e-06]],
      dtype=float32)

In [38]:
val_probs

array([[0.004, 0.354, 0.614, 0.028],
       [0.09 , 0.324, 0.07 , 0.516],
       [0.002, 0.12 , 0.758, 0.12 ],
       ...,
       [0.004, 0.138, 0.738, 0.12 ],
       [0.002, 0.62 , 0.31 , 0.068],
       [0.004, 0.892, 0.07 , 0.034]])

In [264]:
weighted_ans = (val_probs * weights) #for XGBoost
weighted_ans

array([[-0.04  , -0.0354,  0.0614,  0.014 ],
       [-0.9   , -0.0324,  0.007 ,  0.258 ],
       [-0.02  , -0.012 ,  0.0758,  0.06  ],
       ...,
       [-0.04  , -0.0138,  0.0738,  0.06  ],
       [-0.02  , -0.062 ,  0.031 ,  0.034 ],
       [-0.04  , -0.0892,  0.007 ,  0.017 ]])

In [40]:
transformed_weighted_1 = transform_pr1(weighted_ans)
print(transformed_weighted_1)

[1. 1. 1. ... 1. 1. 1.]


In [41]:
res_wei_1 = sum(transformed_weighted_1)/sum(y_val_new)
print( sum(transformed_weighted_1), sum(y_val_new),res_wei_1)
print(accuracy_score(transformed_weighted_1, y_val_new))

171963.0 47081 3.652492512903294
0.6368928199670859


In [117]:
def transform_pr2(predictions): 
    good = predictions[:,2] + predictions[:,3]
    bad = predictions[:,0] + predictions[:,1]
    result = good - bad
  
    result[result >= 0.3] = 1
    result[result < -0.3] = -1
    return result

def back_transform(row):
    if row['score'] > 0.3:
        return 4
    if row['score'] >0 and row['score'] <=0.3:
        return 3
    if row['score'] >= -0.3 and row['score'] <=0:
        return 2
    if row['score'] < -0.3:
        return 1

    
transformed_2 = transform_pr2(v_probs)
transformed_wei_2 = transform_pr2(weighted_ans)

In [159]:

def post_probs_cat(probs):
    weights = np.array([[-1, -0.1, 0.1, 0.5]])
    probs = np.sum(probs * weights, axis=1)
    
    return probs
prob_ans_no_reg = post_probs_cat(v_probs) #when used val_probs as XGBoost probabilities, the accuracy was roughly 56%
prob_ans_no_reg

array([ 0.02346807,  0.03379155,  0.11911843, ...,  0.09998093,
       -0.0946869 , -0.09999931])

def post_probs_cat_1(probs):
    weights = np.array([[-10, -0.1, 0.1, 0.5]])/10.7
    
    probs = np.sum(probs * weights, axis=1)
    
    return probs
prob_ans = post_probs_cat_1(val_probs) 
prob_ans[:20]

In [71]:
import copy
y_val_copy = copy.deepcopy(y_val)
y_val.head(10)

0    3
1    2
2    3
3    4
4    4
5    1
6    3
7    4
8    2
9    3
Name: target, dtype: int64

In [100]:
from sklearn.metrics import mean_absolute_error as mae

In [213]:

def postprocess_probs(probs, thresh):
    weights = np.array([[-10, -0.1, 0.1, 0.5]])
    probs = np.sum(probs * weights, axis=1)
    binarized = probs > thresh
    binarized = binarized * 2 - 1
    return binarized

def score(probs, y_true, thresh):
    answers = postprocess_probs(probs, thresh)
    id2w = id_to_w()
    true_weights = np.array(list(map(id2w.get, y_true)))
    our_score = (true_weights * answers).sum()
    best_score = np.abs(true_weights).sum()
    
    return our_score / best_score, our_score, best_score

sc = score(val_probs, y_val, 0)
sc

(0.4436183395291202, 22410.800000000003, 50518.200000000004)

In [300]:
weights = np.array([[-10, -0.1, 0.1, 0.5]]) #веса, которые надо использовать в метрике, 
#я использовала и для вычисления вероятности. То есть, каждую из четырех вероятностей умножила
#на соответствующий вес - дислайк - на (-10), и т.д. Итоговая сумма в строке - вероятность, которая 
#поможет сформировать бинаризованный ответ (-1 для дислайка и скипа, 1 - для просмотра и лайка.

probs = np.sum(val_probs * weights, axis=1) #v_probs - это вероятности, полученные после XGBoost. 
                                          #val_probs - вероятности после запуска RFC/   
probs


array([ 8.67361738e-18, -6.67400000e-01,  1.03800000e-01, ...,
        8.00000000e-02, -1.70000000e-02, -1.05200000e-01])

По поводу весов: они изрядно улучшают качество конечного результата. Я пробовала менять вес для дислайка и скипа, сначала это были крупные комбинации - -100 и -1, потом путем проб и ошибок вернулась к -0,1 для скипа и -10 для дислайка. Если выбирать значения, близкие к -10, общий результат стабильно ниже на 100-200 от общей суммы.

In [301]:
#Бинаризую ответ. Если вероятность выше нуля - будет "1", ниже нуля - "-1".
thresh = 0

binarized = probs > thresh
binarized = binarized * 2 - 1
binarized[:100]

array([ 1, -1,  1, -1,  1, -1,  1,  1, -1, -1,  1,  1, -1, -1,  1,  1,  1,
        1,  1,  1, -1, -1, -1,  1,  1,  1, -1,  1,  1, -1,  1, -1,  1,  1,
        1, -1,  1,  1,  1,  1,  1,  1, -1,  1,  1,  1,  1,  1,  1,  1, -1,
        1,  1,  1,  1,  1,  1,  1, -1,  1, -1,  1, -1, -1,  1,  1,  1,  1,
        1, -1,  1, -1, -1, -1, -1, -1,  1,  1,  1,  1,  1,  1,  1,  1, -1,
        1, -1, -1,  1,  1, -1, -1,  1, -1, -1,  1,  1,  1, -1, -1])

In [279]:
def id_to_w():
    return {
        1: -10,
        2: -0.1,
        3: 0.1,
        4: 0.5,
    }
id2w = id_to_w()


true_weights = np.array(list(map(id2w.get, y_val)))
true_weights[:50]

array([  0.1,  -0.1,   0.1,   0.5,   0.5, -10. ,   0.1,   0.5,  -0.1,
         0.1,   0.1,   0.5,   0.5,   0.1,  -0.1,   0.5,   0.5,   0.1,
         0.1,   0.5,   0.1,   0.1,  -0.1,   0.1,   0.5,   0.1,   0.1,
         0.1,   0.1,   0.5,   0.5,   0.1,   0.1,   0.1,   0.1,   0.1,
         0.1,   0.5,   0.1,   0.1,   0.1,   0.5,   0.5,   0.1,  -0.1,
        -0.1,   0.1,   0.1,  -0.1,   0.5])

в функции выше назначается вес каждого ответа (dislike - (-10), skip - (-0,1), view - (0.1), like - (0,5))
Дальше переведу ответы в вектор, то есть, вместо 1, 2, 3, 4 - будут стоять веса. Они помогут 
найти сумму реального ответа с учетом веса, где дислайк будет давать 10 баллов, лайк - 0,5 балла, 
скип - 0,1 балла и вью - тоже 0,1 балла.
Не страшно, что некоторые значения со знаком "минус". Сумма будет идти по модулю. Потому что если реальный ответ есть "-1" и вес такого ответа "-10", то получится "10". 

Итого, лучший возможный результат - 50518,2.

In [280]:
best_score = np.abs(true_weights).sum()
best_score

50518.200000000004

Теперь надо посчитать результат, который выдала модель с учетом весов. Для этого возьму готовый вектор с возможными ответами binarized, там все ответы -1 или 1. Если умножить ответ "-1" на вес "-10" в случае правильного дислайка, можно получить +10, что хорошо отразится на общей сумме, но в случае неверного предсказания, результат будет -10. Пан или пропал.

Есть два вектора - один binarized, второй - true_weights. В нём уже собраны веса правильных ответов для валидационного датасета. Умножу их. Получу: 

In [302]:
our_weights = (true_weights * binarized)
our_weights[:200]

array([  0.1,   0.1,   0.1,  -0.5,   0.5,  10. ,   0.1,   0.5,   0.1,
        -0.1,   0.1,   0.5,  -0.5,  -0.1,  -0.1,   0.5,   0.5,   0.1,
         0.1,   0.5,  -0.1,  -0.1,   0.1,   0.1,   0.5,   0.1,  -0.1,
         0.1,   0.1,  -0.5,   0.5,  -0.1,   0.1,   0.1,   0.1,  -0.1,
         0.1,   0.5,   0.1,   0.1,   0.1,   0.5,  -0.5,   0.1,  -0.1,
        -0.1,   0.1,   0.1,  -0.1,   0.5,  -0.1,   0.5,   0.1,   0.1,
         0.1,   0.5,   0.5,   0.1,  -0.1,   0.1,  -0.5,   0.5,  -0.5,
        -0.5,   0.1,   0.1,   0.1,   0.1,   0.1,  -0.5,   0.5,  -0.5,
        -0.1,  -0.5,  -0.5,   0.1,  -0.1,   0.1,   0.1,   0.5,   0.5,
         0.1,  -0.1,   0.1,  -0.5,   0.5,  -0.1,  -0.1,   0.1,  -0.1,
        -0.5,  -0.5,  -0.1,  -0.1,  -0.1,   0.5,   0.1,   0.1,  -0.5,
        -0.5,  -0.5,  -0.1,  10. ,   0.1,   0.5,   0.1,  -0.1,   0.1,
         0.1,   0.5,  -0.1,   0.1,   0.5,  -0.5,  -0.5,   0.1,   0.1,
        -0.1,  -0.1,   0.1,  -0.1,   0.1,   0.1,   0.5,   0.1,   0.1,
        -0.5,   0.5,

Как видно на семпле из 200 ответов, больше неправильно классифицированнх скипов и просмотров, но есть и непойманные лайки, а также один из трех непойманных дислайков. Именно непойманные дислайки будут портить общую сумму. Найду её:

Результаты:
XGBoost - 16487.0

RFC - 22410.800000000003

In [303]:
our_score = our_weights.sum()
our_score

22410.800000000003

Нормализую полученный от Рендом Форест результат на максимально возможную сумму.

In [233]:
our_score/best_score             #  22410.8 / 50518,2

0.4436183395291202

In [265]:
#Попробую другие варианты бинаризованного ответа. Например, разницу между "положительными" и "отрицательными" 
#предсказаниями. Она сохранена в векторе transformed_1

our_weights_diff = (true_weights * transformed_1)
our_weights_diff.sum()

#Результат, мягко говоря, не очень. 7452.399999999998 - XGBoost
                                #   6003.000000000001 - RFC

6003.000000000001

In [262]:
#А что, если попробовать тот же сценарий, но усилить вероятности весами?
#Такой вариант собран в векторе transformed_weighted_1
our_weights_dw = (true_weights * transformed_weighted_1)
our_weights_dw.sum()

#А здесь ещё "лучше": -11331.400000000001 XGBoost

-11331.400000000001

А что нам скажет mean absolute error? С учетом того, что максимально возможная разница между ними "-2", то полученная ошибка 0,704 не так уж и плоха. Что интересно - у XGBoost эта ошибка - 0,603, но эта модель поймала меньше дислайков,

In [240]:
mae_res = mae(y_val_new, binarized)
mae_res

0.6036414810162651

In [270]:
#Нельзя не попробовать mae и для варианта с разницей "положительных и отрицательных" -)))

mae_di = mae(y_val_new, transformed_1)
print(mae_di)

mae_d = mae(y_val_new, transformed_weighted_1)
print(mae_d)

0.5597250571343836
0.7262143600658281


Итог для MAE: если бы нас интересовала только точность классификации без учета веса ответа, тогда вариант с разницей между положительными и отрицательными предсказаниями был бы самым лучшим для сборки ответа, ведь он показывает самую меньшую ошибку. Но так как вес дислайка очень важен, этот вариант не подходит. И в целом МАЕ дает противоположные результаты заданной метрике.

Сборка ответа. Возьму вектор binarized и сделаю из него необходимый датафрейм ('answer_id', 'score')

In [274]:
df_tr = pd.DataFrame(binarized)
df_tr.columns=['score']
df_tr.index.name = 'answer_id'
df_tr.head(10)

score
answer_id       
0              1
1             -1
2              1
3             -1
4              1
5             -1
6              1
7              1
8             -1
9             -1

For fun, добавлю в этот датафрейм реальные ответы, взглянем на первых 50, например.

In [275]:
df_tr['real_answers'] = y_val_new
df_tr.head(50)

score  real_answers
answer_id                     
0              1             1
1             -1            -1
2              1             1
3             -1             1
4              1             1
5             -1            -1
6              1             1
7              1             1
8             -1            -1
9             -1             1
10             1             1
11             1             1
12            -1             1
13            -1             1
14             1            -1
15             1             1
16             1             1
17             1             1
18             1             1
19             1             1
20            -1             1
21            -1             1
22            -1            -1
23             1             1
24             1             1
25             1             1
26            -1             1
27             1             1
28             1             1
29            -1             1
30             1             1
31            -1             1
32             1             1
33             1             1
34             1             1
35            -1             1
36             1             1
37             1             1
38             1             1
39             1             1
40             1             1
41             1             1
42            -1             1
43             1             1
44             1            -1
45             1            -1
46             1             1
47             1             1
48             1            -1
49             1             1

In [276]:
df_tr.drop(['real_answers'], axis=1, inplace=True)
df_tr.to_csv("submission.csv", sep=',', index=True)